sqlx-postgres/src/types/cube.rs

```diff
@@ -0,0 +1,393 @@
+        }
+    }
+}
+
+impl<'q> Encode<'q, Postgres> for PgCube {
+    fn produces(&self) -> Option<PgTypeInfo> {
+        Some(PgTypeInfo::with_name("cube"))
+    }
+
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+        buf.extend_from_slice(&self.serialize());
+
+        IsNull::No
+    }
+}
+
+impl TryFrom<&str> for PgCube {
+    type Error = Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        let content = &input.get(1..input.len() - 1).ok_or(Error::Decode(
+            format!("Could not decode cube string: {}", input).into(),
+        ))?;
+
+        if !content.contains('(') && !content.contains(',') {
+            return parse_point(content);
+        }
+
+        if !content.contains("),(") {
+            return parse_zero_volume(content);
+        }
+
+        let point_vecs = content.split("),(").collect::<Vec<&str>>();
+        if point_vecs.len() == 2 && !point_vecs.iter().any(|pv| pv.contains(',')) {
+            return parse_one_dimensional_interval(point_vecs);
+        }
+
+        parse_multidimensional_interval(point_vecs)
+    }
+}
+
+impl TryFrom<&[u8]> for PgCube {
+    type Error = Error;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let cube_type = bytes
+            .get(0)
+            .map(|&byte| byte as usize)
+            .ok_or(Error::Decode(
+                format!("Could not decode cube bytes: {:?}", bytes).into(),
+            ))?;
+
+        let dimensionality = bytes
+            .get(DIMENSIONALITY_POSITION)
+            .map(|&byte| byte as usize)
+            .ok_or(Error::Decode(
+                format!("Could not decode cube bytes: {:?}", bytes).into(),
+            ))?;
+
+        match (cube_type, dimensionality) {
+            (CUBE_TYPE_ZERO_VOLUME, CUBE_DIMENSION_ONE) => {
+                let point = get_f64_from_bytes(&bytes, 4)?;
+                Ok(PgCube::Point(point))
+            }
+            (CUBE_TYPE_ZERO_VOLUME, _) => {
+                Ok(PgCube::ZeroVolume(deserialize_vector(&bytes, START_INDEX)?))
+            }
+            (CUBE_TYPE_DEFAULT, CUBE_DIMENSION_ONE) => {
+                let x_start = 4;
+                let y_start = x_start + BYTE_WIDTH;
+                let x = get_f64_from_bytes(&bytes, x_start)?;
+                let y = get_f64_from_bytes(&bytes, y_start)?;
+                Ok(PgCube::OneDimensionInterval(x, y))
+            }
+            (CUBE_TYPE_DEFAULT, dim) => Ok(PgCube::MultiDimension(deserialize_matrix(
+                &bytes,
+                START_INDEX,
+                dim,
+            )?)),
+            (flag, dimension) => Err(Error::Decode(
+                format!(
+                    "Could not deserialise cube with flag {} and dimension {}: {:?}",
+                    flag, dimension, bytes
+                )
+                .into(),
+            )),
+        }
+    }
+}
+
+impl PgCube {
+    fn serialize(&self) -> Vec<u8> {
```

Comment:
It's redundant to serialize this to a `Vec` just to copy it to the buffer in the `Encode` impl. This should take a `&mut Vec<u8>` as an argument.

Reply:
For testing convenience, you could add a method like:

```rust
#[cfg(test)] 
fn serialize_to_vec(&self) -> Vec<u8> {
    let mut buff = Vec::new();
    self.serialize(&mut buff);
    buff
}
```